packages/@aws-cdk/lambda-layer-awscli/lib/awscli-layer.ts

```diff
@@ -1,19 +1,150 @@
+    const downloadPath = path.join(downloadDir, `${AwsCliLayer.assetPackageNpmTarPrefix}${targetVersion}.tgz`);
+
+    if (fs.existsSync(downloadPath)) {
+      logs.push(`Using package archive already available at location: ${downloadPath}`);
+      return downloadPath;
+    }
+    logs.push(`Downloading package using npm pack to: ${downloadDir}`);
+    childproc.execSync(`mkdir -p ${downloadDir}; cd ${downloadDir}; npm pack ${AwsCliLayer.assetPackageName}@${targetVersion} -q`);
+    if (fs.existsSync(downloadPath)) {
+      logs.push('Successfully downloaded using npm pack.');
+      return downloadPath;
+    }
+
+    logs.push('Failed to download using npm pack.');
+    return undefined;
+  }
+
+  private static readonly assetPackageName: string = '@aws-cdk/asset-awscli-v1';
+  private static readonly assetPackageNpmTarPrefix: string = 'aws-cdk-asset-awscli-v1-';
+
+  private static requireWrapper(id: string, logs: string[]): any {
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-require-imports
+      const result = require(id);
+      if (result) {
+        logs.push(`require('${id}') succeeded.`);
+      }
+      return result;
+    } catch (e) {
+      logs.push(`require('${id}') failed.`);
+      const eAsError = e as Error;
+      if (eAsError?.stack) {
+        logs.push(eAsError.stack);
+      }
+    }
+  }
+
+  private static installAndLoadPackage(from: string, logs: string[]): any {
+    const installDir = AwsCliLayer.findInstallDir();
+    if (!installDir) {
+      logs.push('Unable to find an install directory. require.main.paths[0] is undefined.');
+      return;
+    }
+    logs.push(`Installing from: ${from} to: ${installDir}`);
+    childproc.execSync(`npm install ${from} --no-save --prefix ${installDir} -q`);
+    return AwsCliLayer.requireWrapper(path.join(installDir, 'node_modules', AwsCliLayer.assetPackageName, 'lib/index.js'), logs);
+  }
+
+  private static findInstallDir(): string | undefined {
+    if (!require.main?.paths) {
+      return undefined;
+    }
+    return require.main.paths[0];
+  }
+
   constructor(scope: Construct, id: string) {
-    super(scope, id, {
-      code: lambda.Code.fromAsset(path.join(__dirname, 'layer.zip'), {
+    const logs: string[] = [];
+    let fallback = false;
+
+    const targetVersion = AwsCliLayer.requireWrapper(path.join(__dirname, '../package.json'), logs).devDependencies[AwsCliLayer.assetPackageName];
+
+    let assetPackage = AwsCliLayer._tryLoadPackage(targetVersion, logs);
+
+    if (!assetPackage) {
+      const downloadPath = AwsCliLayer._downloadPackage(targetVersion, logs);
+      if (downloadPath) {
+        assetPackage = AwsCliLayer.installAndLoadPackage(downloadPath, logs);
+      }
+    }
+
+    let code;
+    if (assetPackage) {
+      // ask for feedback here
+      const asset = new assetPackage.AwsCliAsset(scope, `${id}-asset`);
+      code = lambda.Code.fromBucket(asset.bucket, asset.s3ObjectKey);
+    }
```

Comment:
After successfully loading the package, is there a good way to verify that `AwsCliAsset` is actually present as an API on the package? And that it is of type `s3_assets.Asset`? This is where my TS/JS skills are a little bit lacking. I'm not sure the right way to "type-check" here. Or is it overkill?

Reply:
It all happens at runtime in JavaScript, which doesn't have types. So: we can check for the presence of the symbol, but never its type.

```ts
if (!(assetPackage as any).AwsCliAsset) {
  throw new Error('Ruh roh');
}
```

This is probably good enough, though we might need to stick an API version somewhere...

Reply:
Thank you! What do you mean by API version? The version of `AwsCliAsset` being used will be locked in by the pinned devDependency.